src/ir/common/datalog_ir.h

```diff
@@ -0,0 +1,127 @@
+// /* Copyright 2022 Google LLC.
+//  * Licensed under the Apache License, Version 2.0 (the "License");
+//  * you may not use this file except in compliance with the License.
+//  * You may obtain a copy of the License at
+//  *
+//  *     http://www.apache.org/licenses/LICENSE-2.0
+//  *
+//  * Unless required by applicable law or agreed to in writing, software
+//  * distributed under the License is distributed on an "AS IS" BASIS,
+//  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  * See the License for the specific language governing permissions and
+//  * limitations under the License.
+//  */
+
+// This file contains the datalog IR (DLIR) which makes the translation from
+// this authorization logic into datalog simpler.
+
+#ifndef SRC_IR_COMMON_DATALOG_IR_H_
+#define SRC_IR_COMMON_DATALOG_IR_H_
+
+#include <memory>
+#include <string>
+#include <variant>
+#include <vector>
+
+namespace raksha::ir::common {
+
+// Used to represent whether a predicate is negated or not
+enum Sign { kNegated, kPositive };
+
+// Predicate corresponds to a predicate of the form
+// <pred_name>(arg_1, ..., arg_n), and it may or may not be negated
+class Predicate {
```

Comment:
Just FYI: Ultimately, we should split these off into a separate files, but let us do it later.